src/xunit.netcore.extensions/Discoverers/ConditionalTestDiscoverer.cs

```diff
@@ -25,44 +25,14 @@ internal class ConditionalTestDiscoverer
                                                         IEnumerable<IXunitTestCase> testCases,
                                                         object[] conditionArguments)
         {
-            // A null or empty list of conditionMemberNames is treated as "no conditions".
-            // and the test cases will not be skipped.
-            // Example: [ConditionalFact()]
-            if (conditionArguments == null || conditionArguments.Length == 0)
-            {
-                return testCases;
-            }
-
-            string [] conditionMemberNames;
+            Type calleeType = null;
+            string[] conditionMemberNames = null;
 
-            Type calleeType = conditionArguments[0] as Type;
-            if (calleeType != null)
-            {
-                if (conditionArguments.Length < 2)
-                {
-                    // [ConditionalFact(typeof(x))] no provided methods.
-                    return testCases;
-                }
-
-                // [ConditionalFact(typeof(x), "MethodName")]
-                conditionMemberNames = conditionArguments[1] as string[];
-            }
-            else
-            {
-                // [ConditionalFact("MethodName")]
-                conditionMemberNames = conditionArguments[0] as string[];
-            }
-
-            // [ConditionalFact((string[]) null)]
-            int conditionCount = conditionMemberNames == null ? 0 : conditionMemberNames.Count();
-            if (conditionCount == 0)
-            {
-                return testCases;
-            }
+            if (CheckInputToSkipExecution(conditionArguments, ref calleeType, ref conditionMemberNames, testMethod)) return testCases;
 
             MethodInfo testMethodInfo = testMethod.Method.ToRuntimeMethod();
             Type testMethodDeclaringType = testMethodInfo.DeclaringType;
-            List<string> falseConditions = new List<string>(conditionCount);
+            List<string> falseConditions = new List<string>(conditionMemberNames.Count());
 
             foreach (string entry in conditionMemberNames)
             {
```

Comment:
Should we use EvaluateParameter here as well, so that we don't have duplicate logic/code?

Reply:
Here and `EvaluateParameter` both have for-loop to iterate through each condition, but logic inside the loop is different in two aspects:

1. `EvaluateParameter` won't use reflection to peek into values inside the test class.
2. `Discover` returns test cases (format failure string, and add them to `falseConditions` list), while `EvaluateParameter` immediately returns the boolean (hard to bypass the return type).

It's hard to workaround above two points if we want to do refactoring. Any suggestions will be appreciated!